src/frontend.rs

```diff
@@ -72,271 +76,416 @@ impl fmt::Display for Value {
-/// string<-->hash relation of the keys.
 #[derive(Clone, Debug)]
-pub struct SignedPod {
-    pub pod: backend::SignedPod,
-    // `string_key_map` is a hashmap to store the relation between key strings and key hashes
-    // TODO review if maybe store it as <Hash, String>, so that when iterating we can get each
-    // hash respective string
-    string_key_map: HashMap<String, crate::Hash>,
+pub struct SignedPodBuilder {
+    pub params: Params,
+    pub kvs: HashMap<String, Value>,
 }
 
-impl SignedPod {
-    pub fn new(params: &Params, kvs: HashMap<String, Value>) -> Result<Self> {
-        let (hashed_kvs, string_key_map): (
-            HashMap<crate::Hash, backend::Value>,
-            HashMap<String, crate::Hash>,
-        ) = kvs.iter().fold(
-            (HashMap::new(), HashMap::new()),
-            |(mut hashed_kvs, mut key_to_hash), (k, v)| {
-                let h = hash_str(k);
-                hashed_kvs.insert(h, backend::Value::from(v));
-                key_to_hash.insert(k.clone(), h);
-                (hashed_kvs, key_to_hash)
-            },
-        );
-        let pod = backend::SignedPod::new(params, hashed_kvs)?;
-        Ok(Self {
-            pod,
-            string_key_map,
-        })
+impl SignedPodBuilder {
+    pub fn new(params: &Params) -> Self {
+        Self {
+            params: params.clone(),
+            kvs: HashMap::new(),
+        }
     }
-    pub fn id(&self) -> PodId {
-        self.pod.id
+
+    pub fn insert(&mut self, key: impl Into<String>, value: impl Into<Value>) {
+        self.kvs.insert(key.into(), value.into());
     }
-    pub fn origin(&self) -> Origin {
-        Origin(PodType::Signed, self.pod.id)
+
+    pub fn sign<S: PodSigner>(&self, signer: &mut S) -> SignedPod {
+        let mut kvs = HashMap::new();
+        let mut key_string_map = HashMap::new();
+        for (k, v) in self.kvs.iter() {
+            let k_hash = hash_str(k);
+            kvs.insert(k_hash, middleware::Value::from(v));
+            key_string_map.insert(k_hash, k.clone());
+        }
+        let pod = signer.sign(&self.params, &kvs);
+        SignedPod {
+            pod,
+            key_string_map,
+        }
     }
 }
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub enum NativeStatement {
-    Equal = 2,
-    NotEqual,
-    Gt,
-    Lt,
-    Contains,
-    NotContains,
-    SumOf,
-    ProductOf,
-    MaxOf,
+/// SignedPod is a wrapper on top of backend::SignedPod, which additionally stores the
+/// string<-->hash relation of the keys.
+#[derive(Debug, Clone)]
+pub struct SignedPod {
+    pub pod: Box<dyn middleware::SignedPod>,
+    /// HashMap to store the reverse relation between key strings and key hashes
+    pub key_string_map: HashMap<Hash, String>,
 }
 
-impl From<NativeStatement> for backend::NativeStatement {
-    fn from(v: NativeStatement) -> Self {
-        Self::from_repr(v as usize).unwrap()
+impl SignedPod {
```

Comment:
How would you see merging the `SignedPodBuilder` into `SignedPod`?
where the `SignedPodBuilder::new()` and `SignedPodBuilder.sign()` methods would be just done as part of `SignedPod::new()`.
(and getting rid of the `.insert()` method, which introduces cases where an already signed pod gets new keyvalues inserted)

In this way the interface gets more direct to create pods, with just
```rust
pub struct SignedPod {
    pub fn new<S: PodSigner>(
        params: &Params,
        signer: &mut S,
        kvs: HashMap<String, Value>
    ) -> Self;

    // and the already existing methods
    // id(), origin(), verify(), kvs()
```

Reply:
Please see my comment https://github.com/0xPARC/pod2/pull/18#discussion_r1940820395